gcd/src/main/java/org/spine3/server/storage/datastore/DatastoreWrapper.java

```diff
@@ -20,220 +20,246 @@
-import static org.spine3.protobuf.Messages.fromAny;
-import static org.spine3.protobuf.Messages.toAny;
+import static com.google.common.base.Preconditions.checkState;
 
 /**
- * The Google App Engine Cloud {@link Datastore} wrapper.
+ * Represents a wrapper above GAE {@link Datastore}.
+ * <p>Provides API for datastore to be used in storages.
  *
- * @author Alexander Litus
- * @author Mikhail Mikhaylov
- * @see DatastoreStorageFactory
- * @see LocalDatastoreStorageFactory
+ * @author Dmytro Dashenkov
  */
-/* package */ class DatastoreWrapper {
+/*package*/ class DatastoreWrapper {
 
-    private static final String VALUE_PROPERTY_NAME = "value";
+    private static final String ACTIVE_TRANSACTION_CONDITION_MESSAGE = "Transaction should be active.";
+    private static final String NOT_ACTIVE_TRANSACTION_CONDITION_MESSAGE = "Transaction should NOT be active.";
+
+    private static final Map<String, KeyFactory> keyFactories = new HashMap<>();
 
     private final Datastore datastore;
-    private final DatastoreStorageFactory.Options datastoreOptions;
+    private Transaction activeTransaction;
+    private DatastoreReaderWriter actor;
 
-    /**
-     * Creates a new storage instance.
-     *
-     * @param datastore the datastore implementation to use.
-     */
-    /* package */ DatastoreWrapper(Datastore datastore, DatastoreStorageFactory.Options datastoreOptions) {
+    /*package*/ DatastoreWrapper(Datastore datastore) {
         this.datastore = datastore;
-        this.datastoreOptions = datastoreOptions;
+        this.actor = datastore;
     }
 
     /**
-     * Commits the given mutation.
+     * Wraps {@link Datastore} into an instance of {@code DatastoreWrapper} and returns the instance.
      *
-     * @param mutation the mutation to commit.
-     * @throws RuntimeException if {@link Datastore#commit(CommitRequest)} throws a {@link DatastoreException}
+     * @param datastore          {@link Datastore} to wrap.
+     * @return new instance of {@code DatastoreWrapper}
      */
-    private void commit(Mutation mutation) {
-
-        final CommitRequest commitRequest = CommitRequest.newBuilder()
-                .setMode(NON_TRANSACTIONAL)
-                .setMutation(mutation)
-                .build();
-        try {
-            datastore.commit(commitRequest);
-        } catch (DatastoreException e) {
-            propagate(e);
-        }
+    /*package*/ static DatastoreWrapper wrap(Datastore datastore) {
+        return new DatastoreWrapper(datastore);
     }
 
     /**
-     * Commits the given mutation.
+     * Writes new {@link Entity} into the datastore.
      *
-     * @param mutation the mutation to commit.
-     * @throws RuntimeException if {@link Datastore#commit(CommitRequest)} throws a {@link DatastoreException}
+     * @param entity new {@link Entity} to put into datastore
+     * @throws DatastoreException if an {@link Entity} with such {@link Key} already exists.
```

Comment:
I suspect that duplication is not the only reason. What if the remote storage is unreachable (e.g. network problem)? Looks like we should get `DatastoreException` as well.